src/txn/txn.c

```diff
@@ -592,8 +590,29 @@ __wt_txn_commit(WT_SESSION_IMPL *session, const char *cfg[])
 	    __wt_config_gets_def(session, cfg, "commit_timestamp", 0, &cval));
 	if (cval.len != 0) {
 #ifdef HAVE_TIMESTAMPS
-		WT_ERR(__wt_txn_parse_timestamp(
-		    session, "commit", &txn->commit_timestamp, &cval));
+		wt_timestamp_t ts;
+
+		WT_ERR(__wt_txn_parse_timestamp(session, "commit", &ts, &cval));
+		/*
+		 * commit timestamp being set should be newer than the oldest
+		 * timestamp and should move forward in a transaction
+		 */
+		__wt_readlock(session, &txn_global->rwlock);
+		if (txn_global->has_oldest_timestamp && __wt_timestamp_cmp(&ts,
+		    &txn_global->oldest_timestamp) < 0) {
+			__wt_readunlock(session, &txn_global->rwlock);
```

Comment:
It would be better if the code could be structures so there aren't two calls to __wt_readunlock here, since releasing resources in return paths from the middle of a function can be prone to introducing bugs in the future. Something like the structure here https://github.com/wiredtiger/wiredtiger/blob/develop/src/reconcile/rec_write.c#L1811 might work, but the message output means it would require an `if (ret != 0)` as well.

Reply:
done